core/src/main/java/org/infinispan/commands/functional/ReadWriteKeyValueCommand.java

```diff
@@ -144,20 +147,32 @@ public LoadType loadType() {
    @Override
    public String toString() {
       return new StringBuilder(getClass().getSimpleName())
-         .append(" {key=")
-         .append(toStr(key))
-         .append(", value=").append(toStr(value))
-         .append(", prevValue=").append(toStr(prevValue))
-         .append(", prevMetadata=").append(toStr(prevMetadata))
-         .append(", flags=").append(printFlags())
-         .append(", valueMatcher=").append(valueMatcher)
-         .append(", successful=").append(successful)
-         .append("}")
-         .toString();
+            .append(" {key=")
+            .append(toStr(key))
+            .append(", value=").append(toStr(value))
+            .append(", prevValue=").append(toStr(prevValue))
+            .append(", prevMetadata=").append(toStr(prevMetadata))
+            .append(", flags=").append(printFlags())
+            .append(", valueMatcher=").append(valueMatcher)
+            .append(", successful=").append(successful)
+            .append(", keyDataConversion=").append(keyDataConversion)
+            .append(", valueDataConversion=").append(valueDataConversion)
+            .append("}")
+            .toString();
    }
 
    @Override
    public Mutation toMutation(K key) {
-      return new Mutations.ReadWriteWithValue<>(value, f);
+      V valueFromStorage = (V) valueDataConversion.fromStorage(value);
```

Comment:
Is the `valueFromStorage` guaranteed to be serializable? I guess that since normal commands do all conversions on originator, it's not necessary, but I am not sure what do we require with this compatibility 2.0 stuff. @gustavonalle ?
Could you add a test case when the stored value is and the converted value is not?

Reply:
One use case is, a non-serializable pojo (user type) that is encoded as a UTF-8 byte[ ] with some string representation of it.

Reply:
@karesti I think that the fromStorage conversion here is not right, `value` is of user type (it's not converted in ReadWriteMapImpl.eval`)

Reply:
Had we encoded the value in RWMI (and decoded when executing the command) we'd handle the case as well; we'd also keep the invariant that all commands carry only encoded values. At the cost of (potentially) not-needed conversions.
What path should we go, WDYT?